scripts/builtin/xgboost.dml

```diff
@@ -0,0 +1,780 @@
+        offset = ncol(node_queue) - 1
+        M = addOutputRow(M, node, tree_id, R, offset, best_feature_index, 0.0, 0.0)
+      }
+    }
+    else # has no child => must be leaf
+    {
+      if(sml_type == 1) # Regression
+      {
+        output_value = calculateOutputValue(residual_matrix, lambda)
+      }
+      else # Classification
+      {
+        assert(sml_type == 2)
+        output_value = calculateOutputValueClassification(residual_matrix, curr_prediction, lambda)
+      }
+                                    # offset, best_feature_idx, threshold
+      M = addOutputRow(M, node, tree_id, R, 0.0, 0.0, 0.0, output_value)
+    }
+    node_queue_len = ncol(node_queue)
+  }
+}
+
+
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    node: a 1xn matrix (node containing values)
+# OUTPUT:   left: the new left node id
+# OUTPUT:   right: the new right node id
+calculateChildNodeIDs = function(Matrix[Double] node)  return(Matrix[Double] left, Matrix[Double] right) {
+  left = node * 2.0
+  right = node * 2.0 + 1.0
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    M: the nxn output matrix
+# INPUT:    node: the current node
+# INPUT:    R The matrix R which for each feature in X contains the following information
+#					- R[,2]: 1 (scalar feature)
+#				    - R[,1]: 2 (categorical feature)
+# INPUT:    offset: offset to the left child
+# INPUT:    used_col: which feature got used
+# INPUT:
+# INPUT:    threshold: the value at the node we separated it, if leaf => the output value
+# OUTPUT:   new_M: nxn output Matrix
+addOutputRow = function(
+  Matrix[Double] M,
+  Matrix[Double] node,
+  Double tree_id,
+  Matrix[Double] R,
+  Double offset,
+  Double used_col,
+  Double threshold,
+  Double output_value
+) return (Matrix[Double] new_M) {
+
+      current_node = matrix(0, rows=6, cols=1)
+      current_node[1,1] = node[1,1] # node id
+      current_node[2,1] = tree_id
+      current_node[3,1] = offset # offset to left child
+      current_node[4,1] = used_col # features used
+
+      if(used_col == 0.0) { # is a single value or max depth reached => leaf
+         current_node[5,1] = 0 #  0 if leaf node
+         assert(threshold == 0)
+         current_node[6,1] = output_value
+      }
+      else if (as.scalar(R[1, used_col]) == 2.0) {
+        current_node[5, 1] = 2.0 # categorical
+        assert(threshold == 0.0) # should be 0 at categorical
+        current_node[6,1] = threshold
+      }
+      else {
+        current_node[5,1] = 1.0 # 1 if it is scalar
+        current_node[6,1] = threshold # split value ( 0 if leaf or categorical)
+      }
+
+      if (ncol(M) == 0 & nrow(M) == 0) { # first node
+        new_M = current_node
+      } else {
+        new_M = cbind(M, current_node)
```

Comment:
I think you can always just cbind here, if M is empty, then it will just make the new row. the other logic you are optimizing for is done by the compiler, (if one is empty use the other one as output)

Reply:
done in https://github.com/apache/systemds/pull/1334/commits/69b2a042249479107f19f1ed82f930ad1fbe81bc